### PR TITLE
Hometask4 functions

### PR DESCRIPTION
**Pull request by AndriiBakayev**
Made homatask 4 - functions

### CODE REVIEW

File: src/functions/Functions.php [hunk 1]
```diff
@@ -51,14 +62,19 @@ public function sayHelloArgument($arg): string
...
+     * 
      * @throws \InvalidArgumentException
      */
     public function sayHelloArgumentWrapper($arg): string
     {
         // put your code here
-
+        if (!is_scalar($arg) && !is_null($arg)) {
+            throw new \InvalidArgumentException('sayHelloArgument() function accepts only scalar types: number, string or bool. ');
+            return "Hello";//
```
Comment (yderevianko-cartera) [inline src/functions/Functions.php]: please remove return

File: src/functions/Functions.php [hunk 2]
```diff
@@ -84,13 +100,20 @@ public function countArguments(): array
...
-        // put your code here
+        foreach (func_get_args() as $val) {
+            if (!is_string($val)) {
+                throw new \InvalidArgumentException('countArgumentsWrapper() function accepts strings only');
+            }
+        }
+        return $this->countArguments(...func_get_args());
+
     }
 }
```
Comment (yderevianko-cartera) [inline src/functions/Functions.php]: please add empty line at the end of file

File: tests/CountArgumentsTest.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+     * Makes an instance of a class and prepares for test
+     *
+     * @return void
+     */
+    protected function setUp(): void
+    {
+        $this->functions = new functions\Functions();
+    }
+
+
```
Comment (yderevianko-cartera) [inline tests/CountArgumentsTest.php]: please remove extra empty lines

File: tests/CountArgumentsTest.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    /**
+     * PositiveDataProvider
+     *
+     * @return array provides pairs of ingoing / outgoing arrays fot tests
+     */
+    public function positiveDataProvider(): array
+    {
+        return [
+            [[1,2,3], ['argument_count' => 3 , 'argument_values' => [1,2,3]]],
+            [[1],   ['argument_count' => 1 , 'argument_values' => [1]]],
```
Comment (yderevianko-cartera) [inline tests/CountArgumentsTest.php]: please fix spaces

File: tests/CountArgumentsWrapperTest.php [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+     * 
+     * @param $input takes test value from dataprovider? array of function's params
+     * 
+     * @return void
+     * 
+     * @dataProvider negativeDataProvider
+     */
+    public function testCountArgumentsWrapperArrayOfdifferentTypes($input)
+    {
+        try{
```
Comment (yderevianko-cartera) [inline tests/CountArgumentsWrapperTest.php]: please add space after try

File: tests/CountArgumentsWrapperTest.php [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    /**
+     * NegativeDataProvider provides pairs of ingoing / outgoing arrays 
+     * for negative tests for exceptions throwing
+     *
+     * @return array 
+     */
+    public function negativeDataProvider(): array
+    {
+        return [
+            [['1','2',3,'4','5']],
```
Comment (yderevianko-cartera) [inline tests/CountArgumentsWrapperTest.php]: please add spaces after each ,
and use single quotes ''

File: tests/CountArgumentsWrapperTest.php [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+     */
+    public function positiveDataProvider(): array
+    {
+        return [
+            [['1','2','3'], ['argument_count' => 3 , 'argument_values' => ['1','2','3']]],
+            [['1'],   ['argument_count' => 1 , 'argument_values' => ['1']]],
+            [['0'],   ['argument_count' => 1 , 'argument_values' => ['0']]],
+            [[''],   ['argument_count' => 1 , 'argument_values' => ['']]],
+            [[],   ['argument_count' => 0 , 'argument_values' => []]],
+            [["HELLo"],   ['argument_count' => 1 , 'argument_values' => ["HELLo"]]],
```
Comment (yderevianko-cartera) [inline tests/CountArgumentsWrapperTest.php]: please fix spaces for last two lines

File: tests/CountArgumentsWrapperTest.php [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            [["Hello World"],   ['argument_count' => 1 , 'argument_values' => ["Hello World"]]],
+        ];
+    }
+}
```
Comment (yderevianko-cartera) [inline tests/CountArgumentsWrapperTest.php]: please add empty line at the end of file

File: tests/SayHelloArgumentTest.php [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            [0, 'Hello 0'],
+            [-1, 'Hello -1'],      
+            [-1.5, 'Hello -1.5'], //float
+            [true, 'Hello 1'],
+            [false, 'Hello '],
+            [null, 'Hello '], //Null test
+        ];
+    }
+
+}
```
Comment (yderevianko-cartera) [inline tests/SayHelloArgumentTest.php]: please add empty line at the end of file

File: tests/SayHelloArgumentTest.php [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+            [10, 'Hello 10'],
+            [0, 'Hello 0'],
+            [-1, 'Hello -1'],      
+            [-1.5, 'Hello -1.5'], //float
+            [true, 'Hello 1'],
+            [false, 'Hello '],
+            [null, 'Hello '], //Null test
+        ];
+    }
+
```
Comment (yderevianko-cartera) [inline tests/SayHelloArgumentTest.php]: not needed empty line

File: tests/SayHelloArgumentTest.php [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    /**
+     * SetUp() Makes an instance of a class
+     *
+     * @return void
+     */
+    protected function setUp(): void
+    {
+        $this->functions = new functions\Functions();
+    }
+
```
Comment (yderevianko-cartera) [inline tests/SayHelloArgumentTest.php:42]: not needed empty line

File: tests/SayHelloArgumentWrapperTest.php [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+            $this->functions->sayHelloArgumentWrapper([1,2,3,4,5]);
+            $this->fail('An expected exception not thrown');
+        } catch (InvalidArgumentException $exception){
+            $this->assertStringContainsString(
+                'function accepts only scalar',
+                $exception->getMessage()
+            );
+        }
+    }
+
```
Comment (yderevianko-cartera) [inline tests/SayHelloArgumentWrapperTest.php]: not needed empty line

File: tests/SayHelloArgumentWrapperTest.php [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+     */
+    public function positiveDataProvider(): array
+    {
+        return [
+            [true, 'Hello 1'],
+            [false, 'Hello '],
+            [null, 'Hello '], //Null test
+        ];
+    }
+
```
Comment (yderevianko-cartera) [inline tests/SayHelloArgumentWrapperTest.php]: not needed empty line

File: tests/SayHelloArgumentWrapperTest.php [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+}
```
Comment (yderevianko-cartera) [inline tests/SayHelloArgumentWrapperTest.php]: please add empty line at the end of file

File: tests/SayHelloTest.php [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    public function testSayHello()
+    {
+        $this->assertSame(
+            "Hello", 
+            $this->functions->sayHello(),
+            "functions/Functions.php/SayHello does not say \"Hello\""
+        );
+    }
+
+}
```
Comment (yderevianko-cartera) [inline tests/SayHelloTest.php]: please add empty line at the end of file

File: tests/SayHelloTest.php [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+     */
+    public function testSayHello()
+    {
+        $this->assertSame(
+            "Hello", 
+            $this->functions->sayHello(),
+            "functions/Functions.php/SayHello does not say \"Hello\""
+        );
+    }
+
```
Comment (yderevianko-cartera) [inline tests/SayHelloTest.php]: not needed empty line

File: tests/SayHelloTest.php [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    
+    /**
+     * TestSayHello asserts that output of function is really "Hello"
+     *
+     * @return void
+     */
+    public function testSayHello()
+    {
+        $this->assertSame(
+            "Hello",
```
Comment (yderevianko-cartera) [inline tests/SayHelloTest.php]: please use single quotes for strings

**Review by yderevianko-cartera — CHANGES_REQUESTED** [review]

File: src/functions/Functions.php (hunk 1 above)
Comment (AndriiBakayev) [reply src/functions/Functions.php]: Done

File: src/functions/Functions.php (hunk 2 above)
Comment (AndriiBakayev) [reply src/functions/Functions.php]: Done

File: tests/CountArgumentsTest.php (hunk 3 above)
Comment (AndriiBakayev) [reply tests/CountArgumentsTest.php]: Done

File: tests/CountArgumentsTest.php (hunk 4 above)
Comment (AndriiBakayev) [reply tests/CountArgumentsTest.php]: Done

File: tests/CountArgumentsWrapperTest.php (hunk 5 above)
Comment (AndriiBakayev) [reply tests/CountArgumentsWrapperTest.php]: Done

File: tests/SayHelloArgumentWrapperTest.php (hunk 13 above)
Comment (AndriiBakayev) [reply tests/SayHelloArgumentWrapperTest.php]: Done

File: tests/SayHelloArgumentWrapperTest.php (hunk 14 above)
Comment (AndriiBakayev) [reply tests/SayHelloArgumentWrapperTest.php]: Done

File: tests/SayHelloTest.php (hunk 15 above)
Comment (AndriiBakayev) [reply tests/SayHelloTest.php]: Done

File: tests/SayHelloTest.php (hunk 16 above)
Comment (AndriiBakayev) [reply tests/SayHelloTest.php]: Done

File: tests/SayHelloTest.php (hunk 17 above)
Comment (AndriiBakayev) [reply tests/SayHelloTest.php]: Done

File: tests/CountArgumentsWrapperTest.php (hunk 6 above)
Comment (AndriiBakayev) [reply tests/CountArgumentsWrapperTest.php]: Done

File: tests/CountArgumentsWrapperTest.php (hunk 7 above)
Comment (AndriiBakayev) [reply tests/CountArgumentsWrapperTest.php]: Done

File: tests/CountArgumentsWrapperTest.php (hunk 8 above)
Comment (AndriiBakayev) [reply tests/CountArgumentsWrapperTest.php]: Done

File: tests/SayHelloArgumentTest.php (hunk 9 above)
Comment (AndriiBakayev) [reply tests/SayHelloArgumentTest.php]: Done

File: tests/SayHelloArgumentTest.php (hunk 10 above)
Comment (AndriiBakayev) [reply tests/SayHelloArgumentTest.php]: Done

File: tests/SayHelloArgumentTest.php (hunk 11 above)
Comment (AndriiBakayev) [reply tests/SayHelloArgumentTest.php:42]: Done

File: tests/SayHelloArgumentWrapperTest.php (hunk 12 above)
Comment (AndriiBakayev) [reply tests/SayHelloArgumentWrapperTest.php]: Done

**Review by yderevianko-cartera — APPROVED** [review]

File: src/functions/Functions.php [hunk 18]
```diff
@@ -50,7 +50,6 @@ public function sayHello(): string
     {
         return 'Hello';
     }
-
```
Comment (yderevianko-cartera) [inline src/functions/Functions.php:34]: only remove extra empty line, there should aways be one empty line between methods

**Review by yderevianko-cartera — CHANGES_REQUESTED** [review]

**Review by yderevianko-cartera — APPROVED** [review]

**Pull request merged**